src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -269,16 +270,17 @@ public IList<JsonPropertyInfo> Properties
         // Avoids having to perform an expensive cast to JsonTypeInfo<T> to check if there is a Serialize method.
         internal bool HasSerializeHandler { get; private protected set; }
 
-        // Configure would normally have thrown why initializing properties for source gen but type had SerializeHandler
+        // Exception used to throw on deserialization. In some scenarios i.e.:
+        // configure would have thrown while initializing properties for source gen but type had SerializeHandler
         // so it is allowed to be used for fast-path serialization but it will throw if used for metadata-based serialization
-        internal bool MetadataSerializationNotSupported { get; private protected set; }
+        internal Exception? DeserializationException { get; private protected set; }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal void ValidateCanBeUsedForMetadataSerialization()
         {
-            if (MetadataSerializationNotSupported)
+            if (DeserializationException != null)
             {
-                ThrowHelper.ThrowInvalidOperationException_NoMetadataForTypeProperties(Options.TypeInfoResolver, Type);
+                throw DeserializationException;
```

Comment:
A single `JsonTypeInfo` instance can be shared among multiple threads, so throwing the same extension instance in multiple threads concurrently is a recipe for mangled stackframes. Instead of storing an `Exception` this should store relevant state and throw a relevant, freshly created exception instance.

Reply:
good point, will fix

Reply:
I'll revert this change since it's no longer needed anyway